File: ai/file_validator.py
# -*- coding: utf-8 -*-
"""
File Validator for Document Upload Security

This module provides comprehensive file validation to prevent:
- Path traversal attacks
- Malicious file uploads
- DoS via large files
- File type spoofing
- Basic malware pattern detection

SECURITY: This module implements defense-in-depth for file uploads.
"""

import base64
import binascii
import logging
import os
import re
from typing import Optional, Tuple, List
from dataclasses import dataclass
from enum import Enum


# Configure logging
logger = logging.getLogger(__name__)


class ValidationError(Enum):
    """Enumeration of validation error types"""
    INVALID_BASE64 = "Invalid base64 encoding"
    FILE_TOO_LARGE = "File size exceeds maximum"
    INVALID_EXTENSION = "File extension not allowed"
    INVALID_FILENAME = "Invalid filename (potentially malicious)"
    MAGIC_MISMATCH = "File content does not match extension"
    SUSPICIOUS_CONTENT = "File contains suspicious patterns"
    EMPTY_FILE = "File is empty"


@dataclass
class ValidationResult:
    """Result of file validation"""
    valid: bool
    error_type: Optional[ValidationError] = None
    error_message: str = ""
    file_size: int = 0
    detected_extension: str = ""
    safe_filename: str = ""


class FileValidator:
    """
    Comprehensive file validator for secure document uploads.

    SECURITY FEATURES:
    1. Base64 decoding with size limits
    2. Filename sanitization (path traversal, null byte prevention)
    3. Extension whitelist
    4. Magic number verification (content-based file type detection)
    5. Suspicious content pattern detection
    6. Size limits to prevent DoS
    """

    # Configuration
    MAX_FILE_SIZE = 10 * 1024 * 1024  # 10MB
    MIN_FILE_SIZE = 100  # 100 bytes - reject suspiciously small files
    MAX_FILENAME_LENGTH = 255

    # Allowed file extensions with their magic numbers
    ALLOWED_EXTENSIONS = {
        '.txt': [b''],  # Text files have no specific magic number
        '.md': [b''],   # Markdown
        '.json': [b'{', b'['],  # JSON starts with { or [
        '.csv': [b''],  # CSV is text
        '.pdf': [b'%PDF-'],
        '.html': [b'<html', b'<HTML', b'<!DOCTYPE html'],
        '.htm': [b'<html', b'<HTML', b'<!DOCTYPE html'],
        '.xml': [b'<?xml', b'<xml'],
    }

    # Suspicious patterns that may indicate malware or exploit attempts
    SUSPICIOUS_PATTERNS = [
        # Script injection patterns
        rb'<script',
        rb'javascript:',
        rb'vbscript:',
        rb'data:text/html',

        # PowerShell/Cmd patterns (suspicious in documents)
        rb'powershell',
        rb'cmd\.exe',
        rb'/c\s+',

        # Shell patterns
        rb'/bin/',
        rb'/etc/',
        rb'curl\s+',
        rb'wget\s+',

        # Macro patterns (suspicious in office docs)
        rb'AutoOpen',
        rb'AutoClose',
        rb'Document_Open',
        rb'Workbook_Open',

        # Binary executable patterns
        rb'MZ',  # PE/Windows executable
        rb'\x7fELF',  # Linux executable
        rb'CA FE BA BE',  # Mach-O (macOS) - with spaces for readability
    ]

    # Filename sanitization patterns
    NULL_BYTE = '\x00'
    PATH_TRAVERSAL = re.compile(r'\.\.[/\\]')
    INVALID_CHARS = re.compile(r'[<>:"|?*\x00-\x1f]')

    @classmethod
    def validate_base64_content(
        cls,
        content_b64: str,
        filename: str
    ) -> ValidationResult:
        """
        Validate a base64-encoded file content.

        Args:
            content_b64: Base64-encoded file content
            filename: Original filename (for validation)

        Returns:
            ValidationResult with validation status and details
        """
        # Step 1: Validate filename first (before decoding)
        filename_result = cls.validate_filename(filename)
        if not filename_result.valid:
            return filename_result

        # Step 2: Check base64 length before decoding (prevent DoS)
        if len(content_b64) > (cls.MAX_FILE_SIZE * 4 / 3 + 100):
            logger.warning("Base64 content exceeds maximum length",
                          extra={'filename': filename})
            return ValidationResult(
                valid=False,
                error_type=ValidationError.FILE_TOO_LARGE,
                error_message=f"Base64 content is too large"
            )

        if len(content_b64) == 0:
            return ValidationResult(
                valid=False,
                error_type=ValidationError.EMPTY_FILE,
                error_message="File content is empty"
            )

        # Step 3: Decode base64 with error handling
        try:
            file_content = base64.b64decode(content_b64, validate=True)
        except binascii.Error as e:
            logger.warning("Invalid base64 encoding",
                          extra={'filename': filename, 'error': str(e)})
            return ValidationResult(
                valid=False,
                error_type=ValidationError.INVALID_BASE64,
                error_message="Invalid base64 encoding"
            )

        # Step 4: Check file size
        file_size = len(file_content)
        if file_size > cls.MAX_FILE_SIZE:
            logger.warning("File exceeds maximum size",
                          extra={'filename': filename, 'size': file_size})
            return ValidationResult(
                valid=False,
                error_type=ValidationError.FILE_TOO_LARGE,
                error_message=f"File size ({file_size} bytes) exceeds maximum ({cls.MAX_FILE_SIZE})"
            )

        if file_size < cls.MIN_FILE_SIZE:
            logger.warning("File is suspiciously small",
                          extra={'filename': filename, 'size': file_size})
            return ValidationResult(
                valid=False,
                error_type=ValidationError.EMPTY_FILE,
                error_message=f"File is too small (minimum {cls.MIN_FILE_SIZE} bytes)"
            )

        # Step 5: Validate file extension
        ext = cls._get_extension(filename)
        if ext.lower() not in cls.ALLOWED_EXTENSIONS:
            logger.warning("File extension not allowed",
                          extra={'filename': filename, 'extension': ext})
            return ValidationResult(
                valid=False,
                error_type=ValidationError.INVALID_EXTENSION,
                error_message=f"File extension '{ext}' is not allowed"
            )

        # Step 6: Magic number verification (content-based type check)
        if not cls._verify_magic_number(file_content, ext.lower()):
            logger.warning("File content does not match extension",
                          extra={'filename': filename, 'extension': ext})
            return ValidationResult(
                valid=False,
                error_type=ValidationError.MAGIC_MISMATCH,
                error_message=f"File content does not match the '{ext}' extension"
            )

        # Step 7: Check for suspicious content patterns
        suspicious = cls._check_suspicious_patterns(file_content)
        if suspicious:
            logger.warning("File contains suspicious patterns",
                          extra={'filename': filename, 'patterns': suspicious})
            return ValidationResult(
                valid=False,
                error_type=ValidationError.SUSPICIOUS_CONTENT,
                error_message=f"File contains potentially malicious content: {', '.join(suspicious[:3])}"
            )

        # All checks passed
        return ValidationResult(
            valid=True,
            file_size=file_size,
            detected_extension=ext,
            safe_filename=filename_result.safe_filename
        )

    @classmethod
    def validate_filename(cls, filename: str) -> ValidationResult:
        """
        Validate and sanitize a filename.

        Checks for:
        - Null bytes (injection attacks)
        - Path traversal (../..)
        - Invalid characters
        - Excessive length

        Args:
            filename: The filename to validate

        Returns:
            ValidationResult with safe_filename if valid
        """
        if not filename:
            return ValidationResult(
                valid=False,
                error_type=ValidationError.INVALID_FILENAME,
                error_message="Filename cannot be empty"
            )

        # Check for null bytes
        if cls.NULL_BYTE in filename:
            logger.warning("Filename contains null bytes",
                          extra={'filename': filename})
            return ValidationResult(
                valid=False,
                error_type=ValidationError.INVALID_FILENAME,
                error_message="Filename contains invalid characters"
            )

        # Check for path traversal
        if cls.PATH_TRAVERSAL.search(filename):
            logger.warning("Filename contains path traversal patterns",
                          extra={'filename': filename})
            return ValidationResult(
                valid=False,
                error_type=ValidationError.INVALID_FILENAME,
                error_message="Filename contains invalid path sequences"
            )

        # Check for invalid characters
        if cls.INVALID_CHARS.search(filename):
            logger.warning("Filename contains invalid characters",
                          extra={'filename': filename})
            return ValidationResult(
                valid=False,
                error_type=ValidationError.INVALID_FILENAME,
                error_message="Filename contains invalid characters"
            )

        # Check length
        if len(filename) > cls.MAX_FILENAME_LENGTH:
            return ValidationResult(
                valid=False,
                error_type=ValidationError.INVALID_FILENAME,
                error_message=f"Filename exceeds maximum length of {cls.MAX_FILENAME_LENGTH}"
            )

        # Extract just the basename (remove any path components)
        safe_filename = os.path.basename(filename)

        return ValidationResult(
            valid=True,
            safe_filename=safe_filename
        )

    @classmethod
    def _get_extension(cls, filename: str) -> str:
        """Extract file extension from filename."""
        _, ext = os.path.splitext(filename.lower())
        return ext

    @classmethod
    def _verify_magic_number(cls, content: bytes, extension: str) -> bool:
        """
        Verify that file content matches the expected magic number for the extension.

        Args:
            content: The file content as bytes
            extension: The file extension (e.g., '.pdf')

        Returns:
            True if content matches expected type, False otherwise
        """
        allowed_magic = cls.ALLOWED_EXTENSIONS.get(extension, [])
        if not allowed_magic:
            return False

        # Empty magic number list means "accept any content" (for text files)
        if not allowed_magic or allowed_magic == [b'']:
            return True

        # Check if content starts with any of the expected magic numbers
        for magic in allowed_magic:
            if content.startswith(magic):
                return True

        return False

    @classmethod
    def _check_suspicious_patterns(cls, content: bytes) -> List[str]:
        """
        Scan content for suspicious patterns that may indicate malware.

        Args:
            content: File content to scan

        Returns:
            List of suspicious patterns found (empty if none)
        """
        found = []
        content_lower = content.lower()

        for pattern in cls.SUSPICIOUS_PATTERNS:
            pattern_str = pattern.decode('utf-8', errors='ignore')
            if pattern in content_lower or pattern.replace(b' ', b'') in content:
                found.append(pattern_str)

        return found

    @classmethod
    def get_allowed_extensions(cls) -> List[str]:
        """Return list of allowed file extensions."""
        return list(cls.ALLOWED_EXTENSIONS.keys())

    @classmethod
    def configure(cls, **kwargs):
        """
        Configure validation parameters.

        Args:
            max_file_size: Maximum file size in bytes
            min_file_size: Minimum file size in bytes
            allowed_extensions: Dict of extension -> magic numbers
        """
        if 'max_file_size' in kwargs:
            cls.MAX_FILE_SIZE = kwargs['max_file_size']
        if 'min_file_size' in kwargs:
            cls.MIN_FILE_SIZE = kwargs['min_file_size']
        if 'allowed_extensions' in kwargs:
            cls.ALLOWED_EXTENSIONS = kwargs['allowed_extensions']
        logger.info("FileValidator configured",
                   extra={'max_size': cls.MAX_FILE_SIZE,
                         'min_size': cls.MIN_FILE_SIZE,
                         'extensions': list(cls.ALLOWED_EXTENSIONS.keys())})


# Convenience function for quick validation
def validate_upload(content_b64: str, filename: str) -> Tuple[bool, str]:
    """
    Quick validation wrapper for file uploads.

    Args:
        content_b64: Base64-encoded file content
        filename: Original filename

    Returns:
        Tuple of (is_valid, error_message)
    """
    result = FileValidator.validate_base64_content(content_b64, filename)
    if result.valid:
        return True, ""
    return False, result.error_message


# Export for use in main.py
__all__ = ['FileValidator', 'ValidationResult', 'ValidationError', 'validate_upload']

File: internal/validation/file_validator.go
// Package validation provides comprehensive file validation for secure document uploads
package validation

import (
	"bytes"
	"encoding/base64"
	"fmt"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"unicode"
)

// ValidationErrorType represents different types of validation errors
type ValidationErrorType string

const (
	ErrorInvalidBase64    ValidationErrorType = "Invalid base64 encoding"
	ErrorFileTooLarge     ValidationErrorType = "File size exceeds maximum"
	ErrorInvalidExtension ValidationErrorType = "File extension not allowed"
	ErrorInvalidFilename  ValidationErrorType = "Invalid filename (potentially malicious)"
	ErrorMagicMismatch    ValidationErrorType = "File content does not match extension"
	ErrorSuspiciousContent ValidationErrorType = "File contains suspicious patterns"
	ErrorEmptyFile        ValidationErrorType = "File is empty"
)

// ValidationResult represents the result of file validation
type ValidationResult struct {
	Valid             bool                 `json:"valid"`
	ErrorType         ValidationErrorType   `json:"error_type,omitempty"`
	ErrorMessage      string               `json:"error_message,omitempty"`
	FileSize          int                  `json:"file_size,omitempty"`
	DetectedExtension string               `json:"detected_extension,omitempty"`
	SafeFilename      string               `json:"safe_filename,omitempty"`
	Metadata          map[string]interface{} `json:"metadata,omitempty"`
}

// FileValidator provides comprehensive file validation for secure uploads
type FileValidator struct {
	maxFileSize        int
	minFileSize        int
	maxFilenameLength  int
	allowedExtensions  map[string][][]byte
	suspiciousPatterns [][]byte
	pathTraversalRegex  *regexp.Regexp
	invalidCharsRegex    *regexp.Regexp
}

// DefaultConfig returns a validator with default security settings
func DefaultConfig() *FileValidator {
	return &FileValidator{
		maxFileSize:       10 * 1024 * 1024, // 10MB
		minFileSize:       100,               // 100 bytes
		maxFilenameLength: 255,
		allowedExtensions: map[string][][]byte{
			".txt":  {}, // Text files - no magic number
			".md":   {}, // Markdown
			".json": {{byte('{'), byte('[')}},
			".csv":  {}, // CSV
			".pdf":  {{byte('%'), byte('P'), byte('D'), byte('F')}},
			".html": {{byte('<'), byte('h'), byte('t'), byte('m'), byte('l')}, {byte('<'), byte('H'), byte('T'), byte('M'), byte('L')}, {byte('<'), byte('!'), byte('D'), byte('O'), byte('C'), byte('T'), byte('Y'), byte('P'), byte('E'), byte(' ')}},
			".htm":  {{byte('<'), byte('h'), byte('t'), byte('m'), byte('l')}, {byte('<'), byte('H'), byte('T'), byte('M'), byte('L')}},
			".xml":  {{byte('<'), byte('?'), byte('x'), byte('m'), byte('l')}, {byte('<'), byte('x'), byte('m'), byte('l')}},
		},
		suspiciousPatterns: [][]byte{
			// Script injection patterns
			[]byte("<script"),
			[]byte("javascript:"),
			[]byte("vbscript:"),
			[]byte("data:text/html"),

			// PowerShell/Cmd patterns
			[]byte("powershell"),
			[]byte("cmd.exe"),
			[]byte("/c "),
			[]byte("\\c "),

			// Shell patterns
			[]byte("/bin/"),
			[]byte("/etc/"),
			[]byte("curl "),
			[]byte("wget "),

			// Macro patterns
			[]byte("AutoOpen"),
			[]byte("AutoClose"),
			[]byte("Document_Open"),
			[]byte("Workbook_Open"),

			// Binary executable patterns
			[]byte("MZ"),       // PE/Windows executable
			{0x7f, 'E', 'L', 'F'}, // Linux executable
			{'C', 'A', 0xFE, ' ', 'B', 'A', ' ', 'B', 'E'}, // Mach-O
		},
		pathTraversalRegex:  regexp.MustCompile(`\.\.[\/\\]`),
		invalidCharsRegex:    regexp.MustCompile(`[<>:"|?*\x00-\x1f]`),
	}
}

// New creates a new file validator with custom configuration
func New(maxSize, minSize, maxFilenameLen int) *FileValidator {
	return &FileValidator{
		maxFileSize:       maxSize,
		minFileSize:       minSize,
		maxFilenameLength: maxFilenameLen,
		allowedExtensions: make(map[string][][]byte),
		suspiciousPatterns: [][]byte{
			[]byte("<script"),
			[]byte("javascript:"),
			[]byte("powershell"),
			[]byte("cmd.exe"),
		},
		pathTraversalRegex: regexp.MustCompile(`\.\.[\/\\]`),
		invalidCharsRegex:   regexp.MustCompile(`[<>:"|?*\x00-\x1f]`),
	}
}

// ValidateBase64Content validates a base64-encoded file content
func (v *FileValidator) ValidateBase64Content(contentB64, filename string) ValidationResult {
	// Step 1: Validate filename first
	filenameResult := v.ValidateFilename(filename)
	if !filenameResult.Valid {
		return filenameResult
	}

	// Step 2: Check base64 length before decoding
	maxB64Length := (v.maxFileSize * 4 / 3) + 100
	if len(contentB64) > maxB64Length {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorFileTooLarge,
			ErrorMessage: "Base64 content is too large",
		}
	}

	if len(contentB64) == 0 {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorEmptyFile,
			ErrorMessage: "File content is empty",
		}
	}

	// Step 3: Decode base64
	fileContent, err := base64.StdEncoding.DecodeString(contentB64)
	if err != nil {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorInvalidBase64,
			ErrorMessage: "Invalid base64 encoding: " + err.Error(),
		}
	}

	// Step 4: Check file size
	fileSize := len(fileContent)
	if fileSize > v.maxFileSize {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorFileTooLarge,
			ErrorMessage: formatString("File size (%d bytes) exceeds maximum (%d)", fileSize, v.maxFileSize),
		}
	}

	if fileSize < v.minFileSize {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorEmptyFile,
			ErrorMessage: formatString("File is too small (minimum %d bytes)", v.minFileSize),
		}
	}

	// Step 5: Validate file extension
	ext := v.getExtension(filename)
	ext = strings.ToLower(ext)
	if _, ok := v.allowedExtensions[ext]; !ok {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorInvalidExtension,
			ErrorMessage: formatString("File extension '%s' is not allowed", ext),
		}
	}

	// Step 6: Magic number verification
	if !v.verifyMagicNumber(fileContent, ext) {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorMagicMismatch,
			ErrorMessage: formatString("File content does not match the '%s' extension", ext),
		}
	}

	// Step 7: Check for suspicious content patterns
	suspicious := v.checkSuspiciousPatterns(fileContent)
	if len(suspicious) > 0 {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorSuspiciousContent,
			ErrorMessage: formatString("File contains potentially malicious content: %s", strings.Join(suspicious[:3], ", ")),
		}
	}

	// All checks passed
	return ValidationResult{
		Valid:             true,
		FileSize:          fileSize,
		DetectedExtension: ext,
		SafeFilename:      filenameResult.SafeFilename,
		Metadata:          map[string]interface{}{"content_validated": true},
	}
}

// ValidateFilename validates and sanitizes a filename
func (v *FileValidator) ValidateFilename(filename string) ValidationResult {
	if filename == "" {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorInvalidFilename,
			ErrorMessage: "Filename cannot be empty",
		}
	}

	// Check for null bytes
	if strings.Contains(filename, "\x00") {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorInvalidFilename,
			ErrorMessage: "Filename contains null bytes",
		}
	}

	// Check for path traversal
	if v.pathTraversalRegex.MatchString(filename) {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorInvalidFilename,
			ErrorMessage: "Filename contains invalid path sequences",
		}
	}

	// Check for invalid characters
	if v.invalidCharsRegex.MatchString(filename) {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorInvalidFilename,
			ErrorMessage: "Filename contains invalid characters",
		}
	}

	// Check length
	if len(filename) > v.maxFilenameLength {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorInvalidFilename,
			ErrorMessage: formatString("Filename exceeds maximum length of %d", v.maxFilenameLength),
		}
	}

	// Extract just the basename
	safeFilename := filepath.Base(filename)

	// Additional sanitization: remove control characters
	safeFilename = removeControlChars(safeFilename)

	return ValidationResult{
		Valid:        true,
		SafeFilename: safeFilename,
	}
}

// ValidateBytes validates raw file bytes
func (v *FileValidator) ValidateBytes(content []byte, filename string) ValidationResult {
	// Validate filename first
	filenameResult := v.ValidateFilename(filename)
	if !filenameResult.Valid {
		return filenameResult
	}

	fileSize := len(content)

	// Check file size
	if fileSize > v.maxFileSize {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorFileTooLarge,
			ErrorMessage: formatString("File size (%d bytes) exceeds maximum (%d)", fileSize, v.maxFileSize),
		}
	}

	if fileSize < v.minFileSize {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorEmptyFile,
			ErrorMessage: formatString("File is too small (minimum %d bytes)", v.minFileSize),
		}
	}

	// Validate extension
	ext := v.getExtension(filename)
	ext = strings.ToLower(ext)
	if _, ok := v.allowedExtensions[ext]; !ok {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorInvalidExtension,
			ErrorMessage: formatString("File extension '%s' is not allowed", ext),
		}
	}

	// Magic number verification
	if !v.verifyMagicNumber(content, ext) {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorMagicMismatch,
			ErrorMessage: formatString("File content does not match the '%s' extension", ext),
		}
	}

	// Check for suspicious patterns
	suspicious := v.checkSuspiciousPatterns(content)
	if len(suspicious) > 0 {
		return ValidationResult{
			Valid:        false,
			ErrorType:    ErrorSuspiciousContent,
			ErrorMessage: formatString("File contains potentially malicious content: %s", strings.Join(suspicious[:3], ", ")),
		}
	}

	return ValidationResult{
		Valid:             true,
		FileSize:          fileSize,
		DetectedExtension: ext,
		SafeFilename:      filenameResult.SafeFilename,
	}
}

// getExtension extracts file extension from filename
func (v *FileValidator) getExtension(filename string) string {
	ext := filepath.Ext(filename)
	return ext
}

// verifyMagicNumber checks if content matches the expected magic number
func (v *FileValidator) verifyMagicNumber(content []byte, extension string) bool {
	allowedMagic, ok := v.allowedExtensions[extension]
	if !ok {
		return false
	}

	// Empty magic number list means "accept any content"
	if len(allowedMagic) == 0 {
		return true
	}

	// Check if content starts with any expected magic number
	for _, magic := range allowedMagic {
		if bytes.HasPrefix(content, magic) {
			return true
		}
	}

	return false
}

// checkSuspiciousPatterns scans content for suspicious patterns
func (v *FileValidator) checkSuspiciousPatterns(content []byte) []string {
	contentLower := bytes.ToLower(content)
	found := make([]string, 0)

	for _, pattern := range v.suspiciousPatterns {
		if bytes.Contains(contentLower, pattern) || bytes.Contains(content, pattern) {
			found = append(found, string(pattern))
		}
	}

	return found
}

// AddAllowedExtension adds an allowed file extension with its magic number
func (v *FileValidator) AddAllowedExtension(ext string, magicNumbers ...[]byte) {
	ext = strings.ToLower(ext)
	if ext == "" {
		return
	}
	if !strings.HasPrefix(ext, ".") {
		ext = "." + ext
	}

	if len(magicNumbers) > 0 && magicNumbers[0] != nil {
		v.allowedExtensions[ext] = magicNumbers
	} else {
		v.allowedExtensions[ext] = [][]byte{}
	}
}

// RemoveAllowedExtension removes an allowed file extension
func (v *FileValidator) RemoveAllowedExtension(ext string) {
	ext = strings.ToLower(ext)
	if !strings.HasPrefix(ext, ".") {
		ext = "." + ext
	}
	delete(v.allowedExtensions, ext)
}

// AddSuspiciousPattern adds a suspicious pattern to detect
func (v *FileValidator) AddSuspiciousPattern(pattern string) {
	v.suspiciousPatterns = append(v.suspiciousPatterns, []byte(pattern))
}

// SetSizeLimits sets custom size limits
func (v *FileValidator) SetSizeLimits(minSize, maxSize int) {
	v.minFileSize = minSize
	v.maxFileSize = maxSize
}

// GetAllowedExtensions returns list of allowed extensions
func (v *FileValidator) GetAllowedExtensions() []string {
	exts := make([]string, 0, len(v.allowedExtensions))
	for ext := range v.allowedExtensions {
		exts = append(exts, ext)
	}
	return exts
}

// removeControlChars removes control characters from a string
func removeControlChars(s string) string {
	var builder strings.Builder
	for _, r := range s {
		if !unicode.IsControl(r) || r == '\t' || r == '\n' || r == '\r' {
			builder.WriteRune(r)
		}
	}
	return builder.String()
}

// formatString formats a string with arguments
func formatString(format string, args ...interface{}) string {
	var builder strings.Builder
	currentArg := 0

	for i := 0; i < len(format); i++ {
		if format[i] == '%' && i+1 < len(format) {
			next := format[i+1]
			if next == 'd' || next == 's' || next == 'v' {
				if currentArg < len(args) {
					switch val := args[currentArg].(type) {
					case int:
						builder.WriteString(strconv.Itoa(val))
					case string:
						builder.WriteString(val)
					default:
						builder.WriteString(fmt.Sprintf("%v", val))
					}
					currentArg++
					i++ // Skip the format specifier
					continue
				}
			}
		}
		builder.WriteByte(format[i])
	}

	return builder.String()
}

// ValidateUpload is a convenience function for quick validation
func ValidateUpload(contentB64, filename string) (bool, string) {
	validator := DefaultConfig()
	result := validator.ValidateBase64Content(contentB64, filename)
	if result.Valid {
		return true, ""
	}
	return false, result.ErrorMessage
}

// SanitizeFilename creates a safe filename from an input string
func SanitizeFilename(filename string) string {
	// Get base name
	filename = filepath.Base(filename)

	// Remove extension
	ext := filepath.Ext(filename)
	name := filename[:len(filename)-len(ext)]

	// Replace invalid chars with underscore
	var builder strings.Builder
	for _, r := range name {
		if r == '.' || r == '-' || r == '_' || unicode.IsLetter(r) || unicode.IsDigit(r) {
			builder.WriteRune(r)
		} else if unicode.IsSpace(r) {
			builder.WriteRune('_')
		}
	}

	sanitized := builder.String()

	// Ensure filename is not empty
	if sanitized == "" {
		sanitized = "file"
	}

	// Re-add extension if it was valid
	if ext != "" {
		sanitized += ext
	}

	return sanitized
}

// IsTextFile checks if the content appears to be a text file
func IsTextFile(content []byte) bool {
	if len(content) == 0 {
		return true
	}

	// Check if content is mostly printable ASCII
	printableCount := 0
	for _, b := range content {
		if b >= 32 && b <= 126 {
			printableCount++
		} else if b == '\t' || b == '\n' || b == '\r' {
			printableCount++
		}
	}

	if float64(printableCount)/float64(len(content)) > 0.9 {
		return true
	}

	return false
}

// GetMimeType returns the MIME type for a given file extension
func GetMimeType(filename string) string {
	ext := strings.ToLower(filepath.Ext(filename))

	mimeTypes := map[string]string{
		".txt":  "text/plain",
		".md":   "text/markdown",
		".json": "application/json",
		".csv":  "text/csv",
		".pdf":  "application/pdf",
		".html": "text/html",
		".htm":  "text/html",
		".xml":  "application/xml",
		".zip":  "application/zip",
		".gz":   "application/gzip",
	}

	if mime, ok := mimeTypes[ext]; ok {
		return mime
	}

	return "application/octet-stream"
}

// GetFileCategory returns the category of a file (document, image, archive, etc.)
func GetFileCategory(filename string) string {
	ext := strings.ToLower(filepath.Ext(filename))

	documentExts := map[string]bool{
		".txt": true, ".md": true, ".json": true, ".csv": true,
		".pdf": true, ".html": true, ".htm": true, ".xml": true,
		".doc": true, ".docx": true, ".xls": true, ".xlsx": true,
		".ppt": true, ".pptx": true,
	}

	imageExts := map[string]bool{
		".jpg": true, ".jpeg": true, ".png": true, ".gif": true,
		".bmp": true, ".svg": true, ".webp": true,
	}

	archiveExts := map[string]bool{
		".zip": true, ".gz": true, ".tar": true, ".rar": true,
		".7z": true, ".bz2": true,
	}

	if documentExts[ext] {
		return "document"
	}
	if imageExts[ext] {
		return "image"
	}
	if archiveExts[ext] {
		return "archive"
	}

	return "unknown"
}
